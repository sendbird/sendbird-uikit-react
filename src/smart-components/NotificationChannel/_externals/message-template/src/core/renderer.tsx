import React from 'react';

import type { BasicProps, ComponentType, ComponentsUnion, GetProperties } from '../types/components';

type FC<Props, ParsedStyle> = (props: BasicProps<Props, ParsedStyle>) => React.ReactElement | null;

type ComponentProperties<Type extends ComponentType> = GetProperties<Type, ComponentsUnion>;

interface RendererOptions<ParsedStyle> {
  views?: {
    [component in ComponentType]?: FC<ComponentProperties<component>, ParsedStyle>;
  };
}

export type Renderer<ParsedStyle> = {
  [component in ComponentType]: FC<ComponentProperties<component>, ParsedStyle>;
};

const FRAGMENT = ({ children }: React.PropsWithChildren) => <React.Fragment>{children}</React.Fragment>;

export function createRenderer<ParsedStyle = object | string>(
  opts?: RendererOptions<ParsedStyle>,
): Renderer<ParsedStyle> {
  return {
    box: opts?.views?.box || FRAGMENT,
    text: opts?.views?.text || FRAGMENT,
    image: opts?.views?.image || FRAGMENT,
    imageButton: opts?.views?.imageButton || FRAGMENT,
    textButton: opts?.views?.textButton || FRAGMENT,
  };
}
