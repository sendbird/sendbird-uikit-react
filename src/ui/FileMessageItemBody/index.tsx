import React, { ReactElement } from 'react';
import { FileMessage } from 'sendbird';
import './index.scss';

import Label, { LabelTypography, LabelColors } from '../Label';
import Icon, { IconTypes, IconColors } from '../Icon';
import TextButton from '../TextButton';
import { getClassName, getUIKitFileType, truncateString } from '../../utils';

interface Props {
  className?: string | Array<string>;
  message: FileMessage;
  isByMe?: boolean;
  mouseHover?: boolean;
}

export default function FileMessageItemBody({
  className,
  message,
  isByMe = false,
  mouseHover = false,
}: Props): ReactElement {

  return (
    <div className={getClassName([
      className,
      'sendbird-file-message-item-body',
      isByMe ? 'outgoing' : 'incoming',
      mouseHover ? 'mouse-hover' : '',
      message?.reactions?.length > 0 ? 'reactions' : '',
    ])}>
      <div className="sendbird-file-message-item-body__file-icon">
        <Icon
          className={'sendbird-file-message-item-body__file-icon__icon'}
          type={{
            IMAGE: IconTypes.PHOTO,
            VIDEO: IconTypes.PLAY,
            AUDIO: IconTypes.FILE_AUDIO,
            GIF: IconTypes.GIF,
            OTHERS: IconTypes.FILE_DOCUMENT,
          }[getUIKitFileType(message?.type)]}
          fillColor={IconColors.PRIMARY}
          width="24px"
          height="24px"
        />
      </div>
      <TextButton
        className="sendbird-file-message-item-body__file-name"
        onClick={() => { window.open(message?.url) }}
        color={isByMe ? LabelColors.ONCONTENT_1 : LabelColors.ONBACKGROUND_1}
      >
        <Label
          className="sendbird-file-message-item-body__file-name__text"
          type={LabelTypography.BODY_1}
          color={isByMe ? LabelColors.ONCONTENT_1 : LabelColors.ONBACKGROUND_1}
        >
          {truncateString(message?.name || message?.url)}
        </Label>
      </TextButton>
    </div>
  );
}
