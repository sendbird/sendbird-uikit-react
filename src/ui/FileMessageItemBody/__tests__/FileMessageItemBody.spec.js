import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import FileMessageItemBody from "../index";

const createMockMessage = (process) => {
  const mockMessage = {
    type: 'image/png',
    url: 'https://sendbird.com',
    name: 'My name is name',
    reactions: [],
  };
  return process ? process(mockMessage) : mockMessage;
};

describe('FileMessageItemBody', () => {
  it('should have class names of own basic status', () => {
    const className = 'classname-for-test';
    const component = mount(
      <FileMessageItemBody
        className={className}
        message={createMockMessage()}
      />
    );
    expect(
      component.find('.sendbird-file-message-item-body').hostNodes().exists()
    ).toBe(true);
    expect(
      component.find(`.${className}.sendbird-file-message-item-body`).hostNodes().exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-file-message-item-body.outgoing').hostNodes().exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-file-message-item-body.incoming').hostNodes().exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-file-message-item-body.mouse-hover').hostNodes().exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-file-message-item-body.reactions').hostNodes().exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-file-message-item-body__file-icon').hostNodes().exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-file-message-item-body__file-icon__icon').hostNodes().exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-file-message-item-body__file-name').hostNodes().exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-file-message-item-body__file-name__text').hostNodes().exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-file-message-item-body__file-name__text').hostNodes().last().text()
    ).toBe(createMockMessage().name);
  });

  it('should render url when name does not exist', () => {
    const component = mount(
      <FileMessageItemBody
        message={createMockMessage((mock) => ({
          ...mock,
          name: '',
        }))}
      />
    );
    expect(
      component.find('.sendbird-file-message-item-body__file-name').hostNodes().exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-file-message-item-body__file-name__text').hostNodes().last().text()
    ).toBe(createMockMessage().url);
  });

  it('should have class name by isByMe prop', () => {
    const outgoingMessage = mount(
      <FileMessageItemBody
        message={createMockMessage()}
        isByMe
      />
    );
    expect(
      outgoingMessage.find('.sendbird-file-message-item-body.outgoing').hostNodes().exists()
    ).toBe(true);
    expect(
      outgoingMessage.find('.sendbird-file-message-item-body.incoming').hostNodes().exists()
    ).toBe(false);
    const incomingMessage = mount(
      <FileMessageItemBody
        message={createMockMessage()}
        isByMe={false}
      />
    );
    expect(
      incomingMessage.find('.sendbird-file-message-item-body.outgoing').hostNodes().exists()
    ).toBe(false);
    expect(
      incomingMessage.find('.sendbird-file-message-item-body.incoming').hostNodes().exists()
    ).toBe(true);
  });

  it('should have class name by mouseHover prop', () => {
    const component = mount(
      <FileMessageItemBody
        message={createMockMessage()}
        mouseHover
      />
    );
    expect(
      component.find('.sendbird-file-message-item-body.mouse-hover').hostNodes().exists()
    ).toBe(true);
  });

  it('should have class name by reactions of message prop', () => {
    const component = mount(
      <FileMessageItemBody
        isReactionEnabled
        message={createMockMessage((mock) => ({
          ...mock,
          reactions: [{}, {}],
        }))}
      />
    );
    expect(
      component.find('.sendbird-file-message-item-body.reactions').hostNodes().exists()
    ).toBe(true);
  });

  it('should render icons by type of message prop', () => {
    const imageMessage = mount(
      <FileMessageItemBody
        message={createMockMessage((mock) => ({
          ...mock,
          type: 'image/jpg',
        }))}
      />
    );
    expect(
      imageMessage.find('.sendbird-file-message-item-body').hostNodes().exists()
    ).toBe(true);
    expect(
      imageMessage.find('.sendbird-icon').hostNodes().exists()
    ).toBe(true);
    expect(
      imageMessage.find('.sendbird-icon-photo').hostNodes().exists()
    ).toBe(true);
    const videoMessage = mount(
      <FileMessageItemBody
        message={createMockMessage((mock) => ({
          ...mock,
          type: 'video/mp4',
        }))}
      />
    );
    expect(
      videoMessage.find('.sendbird-file-message-item-body').hostNodes().exists()
    ).toBe(true);
    expect(
      videoMessage.find('.sendbird-icon').hostNodes().exists()
    ).toBe(true);
    expect(
      videoMessage.find('.sendbird-icon-play').hostNodes().exists()
    ).toBe(true);
    const audioMessage = mount(
      <FileMessageItemBody
        message={createMockMessage((mock) => ({
          ...mock,
          type: 'audio/mp3',
        }))}
      />
    );
    expect(
      audioMessage.find('.sendbird-file-message-item-body').hostNodes().exists()
    ).toBe(true);
    expect(
      audioMessage.find('.sendbird-icon').hostNodes().exists()
    ).toBe(true);
    expect(
      audioMessage.find('.sendbird-icon-file-audio').hostNodes().exists()
    ).toBe(true);
    const gifMessage = mount(
      <FileMessageItemBody
        message={createMockMessage((mock) => ({
          ...mock,
          type: 'image/gif',
        }))}
      />
    );
    expect(
      gifMessage.find('.sendbird-file-message-item-body').hostNodes().exists()
    ).toBe(true);
    expect(
      gifMessage.find('.sendbird-icon').hostNodes().exists()
    ).toBe(true);
    expect(
      gifMessage.find('.sendbird-icon-gif').hostNodes().exists()
    ).toBe(true);
    const documentMessage = mount(
      <FileMessageItemBody
        message={createMockMessage((mock) => ({
          ...mock,
          type: 'image/pdf',
        }))}
      />
    );
    expect(
      documentMessage.find('.sendbird-file-message-item-body').hostNodes().exists()
    ).toBe(true);
    expect(
      documentMessage.find('.sendbird-icon').hostNodes().exists()
    ).toBe(true);
    expect(
      documentMessage.find('.sendbird-icon-file-document').hostNodes().exists()
    ).toBe(true);
  });

  it('should do a snapshot test of the FileMessageItemBody DOM', function () {
    const component = renderer.create(
      <FileMessageItemBody
        className="classname-for-snapshot"
        message={createMockMessage((mock) => ({
          ...mock,
          reactions: [{}, {}],
        }))}
        isByMe
        mouseHover
      />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
