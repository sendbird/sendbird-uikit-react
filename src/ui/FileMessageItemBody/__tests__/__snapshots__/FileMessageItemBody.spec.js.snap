// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui/FileMessageItemBody should do a snapshot test of the FileMessageItemBody DOM 1`] = `
<DocumentFragment>
  <div
    class="classname-for-snapshot sendbird-file-message-item-body outgoing mouse-hover "
  >
    <div
      class="sendbird-file-message-item-body__file-icon"
    >
      <div
        class="sendbird-file-message-item-body__file-icon__icon sendbird-icon sendbird-icon-photo sendbird-icon-color--primary"
        role="button"
        style="width: 24px; min-width: 24px; height: 24px; min-height: 24px;"
        tabindex="0"
      >
        <test-file-stub />
      </div>
    </div>
    <div
      class="sendbird-file-message-item-body__file-name sendbird-color--oncontent-1 sendbird-textbutton "
      role="button"
      tabindex="0"
    >
      <span
        class="sendbird-file-message-item-body__file-name__text sendbird-label sendbird-label--body-1 sendbird-label--color-oncontent-1"
      >
        My name is name
      </span>
    </div>
  </div>
</DocumentFragment>
`;
